File: 프로그래머스/3/70130. 스타 수열/스타 수열.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

int len_starseq(int num, vector<int>& a){
    int result = 0;
    for(int i = 0; i + 1 < a.size(); i++){
        if(a[i] != a[i + 1] && (a[i] == num || a[i + 1] == num)){
            result += 2;
            i++;
        }
    }
    
    return result;
}

int solution(vector<int> a) {
    map<int, int> counts;
    for(int num : a) counts[num]++;
    
    int answer = 0;
    for(auto iter = counts.begin(); iter != counts.end(); iter++){
        if(iter->second * 2 <= answer) continue;
        
        answer = max(answer, len_starseq(iter->first, a));
    }
    
    return answer;
}

File: 프로그래머스/3/70130. 스타 수열/스타 수열.py
from collections import defaultdict

def len_starseq(num, a):
    result = 0
    i = 0
    while i + 1 < len(a):
        if a[i] != a[i + 1] and (a[i] == num or a[i + 1] == num):
            result += 2
            i += 1
        
        i += 1
    
    return result

def solution(a):
    counts = defaultdict(int)
    for num in a:
        counts[num] += 1
    
    answer = 0
    for num in counts.keys():
        if counts[num] * 2 <= answer:
            continue
        
        answer = max(answer, len_starseq(num, a))
    
    return answer